chain/chain/src/chain.rs

```diff
@@ -4275,88 +4340,74 @@ impl Chain {
+        let (execution_contexts_after, _) = self.get_blocks_range_with_missing_chunks(
+            me,
+            *prev_block.hash(),
+            prev_chunk_height_included,
+            shard_id,
+            mode,
+            false,
+        )?;
+
+        // Create stateless validation job.
+        // Its initial state corresponds to the block at which `prev_prev_chunk`
+        // was created, and then it:
+        // 1. processes updates for missing chunks until a block at which
+        // `prev_chunk` was created;
+        // 2. processes update for the `prev_chunk`;
+        // 3. processes updates for missing chunks until the last chunk
+        // is reached.
         Ok(Some((
             shard_id,
             Box::new(move |parent_span| -> Result<ShardUpdateResult, Error> {
-                let mut result = vec![];
-                for (block_context, shard_context) in execution_contexts {
-                    let block_result = process_shard_update(
-                        parent_span,
-                        runtime.as_ref(),
-                        epoch_manager.as_ref(),
-                        ShardUpdateReason::OldChunk(OldChunkData {
-                            block: block_context.clone(),
-                            split_state_roots: None,
-                            prev_chunk_extra: current_chunk_extra.clone(),
-                            storage_context: StorageContext {
-                                storage_data_source: StorageDataSource::DbTrieOnly,
-                                state_patch: Default::default(),
-                            },
-                        }),
-                        shard_context,
-                    )?;
-                    if let ShardBlockUpdateResult::OldChunk(OldChunkResult {
-                        shard_uid,
-                        apply_result,
-                        apply_split_result_or_state_changes: _,
-                    }) = block_result
-                    {
-                        *current_chunk_extra.state_root_mut() = apply_result.new_root;
-                        result.push((
-                            block_context.block_hash,
-                            shard_uid,
-                            current_chunk_extra.clone(),
-                        ));
-                    }
-                }
+                // Process missing chunks before previous chunk.
+                let mut result = process_missing_chunks_range(
+                    parent_span,
+                    &mut current_chunk_extra,
+                    runtime.as_ref(),
+                    epoch_manager.as_ref(),
+                    execution_contexts_before,
+                )?;
                 // TODO(logunov): use `validate_chunk_with_chunk_extra`
                 assert_eq!(current_chunk_extra.state_root(), &prev_chunk.prev_state_root());
+                // Process previous chunk.
                 let block_result = process_shard_update(
```

Comment:
I also suggest making `apply_new_chunk` crate public and calling it here directly instead of  `process_shard_update`